nn/transformer.py

```diff
@@ -0,0 +1,301 @@
+
+    self.num_layers = num_layers
+    self.norm = norm
+    self.norm_eps = norm_eps
+
+  def forward(self, x: nn.LayerRef) -> nn.LayerRef:
+    """
+    Applies every encoder layer initialized in self.layers.
+    """
+    output = x
+
+    for mod in self.layers:
+      output = mod(output)
+
+    if self.norm is not None:
+      output = self.norm(output, epsilon=self.norm_eps)
+
+    return output
+
+
+class TransformerDecoderLayerStep(nn.Module):
+  """
+  Defines one layer of a standard transformer decoder
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, dropout: float = 0.1, att_dropout: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu, norm_eps: float = 1e-5,
+               norm_first: bool = False, norm=nn.layer_norm):
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param att_dropout: dropout value for attention
+    :param activation: activation function
+    :param norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
+    :param norm: normalization function
+    """
+    super().__init__()
+    self.self_attn = nn.SelfAttentionStep(
+      key_dim_total=dim_model, value_dim_total=dim_model, num_heads=num_heads, att_dropout=att_dropout)
+    self.multihead_attn = nn.Attention()
+
+    self.linear_ff = nn.Linear(dim_ff)
+    self.linear_out = nn.Linear(dim_model)
+
+    self.norm = norm
+    self.norm_first = norm_first
+    self.norm_eps = norm_eps
+    self.activation = activation
+    self.dropout = dropout
+    self.att_dropout = att_dropout
+
+  def forward(self, x: nn.LayerRef, *, memory: nn.LayerRef, state: nn.LayerState) -> Tuple[nn.LayerRef, nn.LayerState]:
+    """
+    Two possible forward variants of decoder, defined by self.norm_first, x and memory have shape {B, T, F}
+    """
+    if self.norm_first:
+      x, new_state = x + self._self_attention_block(self.norm(x, epsilon=self.norm_eps), state=state)
+      x = x + self._multi_head_attention_block(self.norm(x, epsilon=self.norm_eps), memory)
+      x = x + self._feed_forward_block(self.norm(x, epsilon=self.norm_eps))
+    else:
+      x, new_state = self.norm(x + self._self_attention_block(x, state=state), epsilon=self.norm_eps)
+      x = self.norm(x + self._multi_head_attention_block(x, memory), epsilon=self.norm_eps)
+      x = self.norm(x + self._feed_forward_block(x), epsilon=self.norm_eps)
+
+    return x, new_state
+
+  def initial_state(self) -> nn.LayerState:
+    """
+    initial state declaration
+    """
+    return nn.LayerState(
+      self_attention=self.self_attn.initial_state(),
```

Comment:
Always check consistency. If the attrib is called `self_attn`, you also should call the state key the same, i.e. `self_attn`.